File: sdks/go/bot.go
package lsmsg

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"runtime"
	"strings"
	"sync"
)

// EventHandler is the callback signature for event handlers.
type EventHandler func(event *Event) error

// HandlerOpts specifies optional filters for handler registration.
type HandlerOpts struct {
	Pattern  string
	Platform Platform
}

// InvokeOpt configures an Invoke call.
type InvokeOpt func(*invokeConfig)

type invokeConfig struct {
	input    map[string]any
	config   map[string]any
	metadata map[string]any
}

// WithInput sets the input for a LangGraph run.
func WithInput(input map[string]any) InvokeOpt {
	return func(c *invokeConfig) { c.input = input }
}

// WithConfig sets the config for a LangGraph run.
func WithConfig(config map[string]any) InvokeOpt {
	return func(c *invokeConfig) { c.config = config }
}

// WithMetadata sets the metadata for a LangGraph run.
func WithMetadata(metadata map[string]any) InvokeOpt {
	return func(c *invokeConfig) { c.metadata = metadata }
}

// BotConfig holds configuration for creating a new Bot.
type BotConfig struct {
	Slack           *SlackConfig
	Teams           *TeamsConfig
	LangGraph       *LangGraphConfig
	MaxPendingTasks int
	MaxBodyBytes    int
}

// SlackConfig holds Slack-specific configuration.
type SlackConfig struct {
	SigningSecret string
	BotToken      string
}

// TeamsConfig holds Teams-specific configuration.
type TeamsConfig struct {
	AppID       string
	AppPassword string
}

// LangGraphConfig holds LangGraph client configuration.
type LangGraphConfig struct {
	URL    string
	APIKey string
}

// Bot is the main entry point for handling messaging platform webhooks.
// It implements http.Handler.
type Bot struct {
	config         BotConfig
	ffi            ffiBackend
	mu             sync.RWMutex
	handlers       map[int64]EventHandler
	registryHandle int64
	closed         bool

	// lgHandle is the FFI handle for the LangGraph client. Handle 0 is valid.
	lgHandle int64

	langGraphConfigured bool
	httpClient          *http.Client
	slackAPIBaseURL     string
}

// defaultBackend is set by the cgo init function. It is nil when the shared
// library is not linked (e.g. in pure-Go unit tests).
var defaultBackend ffiBackend

// NewBot creates a new Bot with the given configuration.
// It panics if the FFI shared library was not linked. Use newBotWithBackend
// in tests to supply a mock backend.
func NewBot(config BotConfig) *Bot {
	if defaultBackend == nil {
		panic("lsmsg: FFI backend not available — build with CGO_ENABLED=1 and the lsmsg_ffi shared library")
	}
	return newBotWithBackend(config, defaultBackend)
}

// newBotWithBackend creates a Bot with a custom FFI backend (for testing).
func newBotWithBackend(config BotConfig, backend ffiBackend) *Bot {
	if config.MaxBodyBytes == 0 {
		config.MaxBodyBytes = 1 << 20
	}
	if config.MaxPendingTasks == 0 {
		config.MaxPendingTasks = 100
	}
	registryHandle := backend.RegistryNew()
	if registryHandle < 0 {
		panic("lsmsg: failed to create handler registry")
	}
	b := &Bot{
		config:              config,
		ffi:                 backend,
		handlers:            make(map[int64]EventHandler),
		registryHandle:      registryHandle,
		langGraphConfigured: config.LangGraph != nil,
		httpClient:          http.DefaultClient,
		slackAPIBaseURL:     "https://slack.com",
	}
	if config.LangGraph != nil {
		b.lgHandle = b.ffi.LangGraphNew(config.LangGraph.URL, config.LangGraph.APIKey)
	}
	runtime.SetFinalizer(b, (*Bot).Close)
	return b
}

// Close releases native resources owned by the bot. It is safe to call multiple times.
func (b *Bot) Close() {
	runtime.SetFinalizer(b, nil)

	b.mu.Lock()
	if b.closed {
		b.mu.Unlock()
		return
	}
	b.closed = true
	registryHandle := b.registryHandle
	lgHandle := b.lgHandle
	freeLangGraph := b.langGraphConfigured
	b.mu.Unlock()

	b.ffi.RegistryFree(registryHandle)
	if freeLangGraph {
		b.ffi.LangGraphFree(lgHandle)
	}
}

func (b *Bot) register(kind EventKind, command, emoji string, platform Platform, rawEventType string, pattern string, handler EventHandler) {
	fields := map[string]string{}
	if kind != "" {
		fields["event_kind"] = string(kind)
	}
	if command != "" {
		fields["command"] = command
	}
	if emoji != "" {
		fields["emoji"] = emoji
	}
	if platform != "" {
		fields["platform"] = string(platform)
	}
	if rawEventType != "" {
		fields["raw_event_type"] = rawEventType
	}
	if pattern != "" {
		fields["pattern"] = pattern
	}

	fieldsJSON, err := json.Marshal(fields)
	if err != nil {
		panic(fmt.Sprintf("lsmsg: failed to marshal handler filter: %v", err))
	}
	id := b.ffi.RegistryRegister(b.registryHandle, fieldsJSON)
	if id < 0 {
		panic("lsmsg: failed to register handler filter")
	}

	b.mu.Lock()
	defer b.mu.Unlock()
	b.handlers[id] = handler
}

// OnMention registers a handler for mention events.
func (b *Bot) OnMention(handler EventHandler) {
	b.register(EventMention, "", "", "", "", "", handler)
}

// OnMentionWithOpts registers a mention handler with additional filters.
func (b *Bot) OnMentionWithOpts(handler EventHandler, opts HandlerOpts) {
	b.register(EventMention, "", "", opts.Platform, "", opts.Pattern, handler)
}

// OnMessage registers a handler for message events.
func (b *Bot) OnMessage(handler EventHandler) {
	b.register(EventMessage, "", "", "", "", "", handler)
}

// OnMessageWithOpts registers a message handler with additional filters.
func (b *Bot) OnMessageWithOpts(handler EventHandler, opts HandlerOpts) {
	b.register(EventMessage, "", "", opts.Platform, "", opts.Pattern, handler)
}

// Command registers a handler for a slash command.
func (b *Bot) Command(name string, handler EventHandler) {
	b.register(EventCommand, name, "", "", "", "", handler)
}

// CommandWithOpts registers a command handler with additional filters.
func (b *Bot) CommandWithOpts(name string, handler EventHandler, opts HandlerOpts) {
	b.register(EventCommand, name, "", opts.Platform, "", opts.Pattern, handler)
}

// OnReaction registers a handler for a specific emoji reaction.
func (b *Bot) OnReaction(emoji string, handler EventHandler) {
	b.register(EventReaction, "", emoji, "", "", "", handler)
}

// OnReactionWithOpts registers a reaction handler with additional filters.
func (b *Bot) OnReactionWithOpts(emoji string, handler EventHandler, opts HandlerOpts) {
	b.register(EventReaction, "", emoji, opts.Platform, "", opts.Pattern, handler)
}

// On registers a handler for a raw event type.
func (b *Bot) On(eventType string, handler EventHandler) {
	b.register(EventRaw, "", "", "", eventType, "", handler)
}

// OnWithOpts registers a raw event handler with additional filters.
func (b *Bot) OnWithOpts(eventType string, handler EventHandler, opts HandlerOpts) {
	b.register(EventRaw, "", "", opts.Platform, eventType, opts.Pattern, handler)
}

func (b *Bot) matchHandlerIDs(event *Event) ([]int64, error) {
	eventJSON, err := json.Marshal(event)
	if err != nil {
		return nil, fmt.Errorf("lsmsg: failed to marshal event: %w", err)
	}
	ids, err := b.ffi.RegistryMatchEvent(b.registryHandle, eventJSON)
	if err != nil {
		return nil, fmt.Errorf("lsmsg: failed to match event: %w", err)
	}
	return ids, nil
}

func (b *Bot) dispatchMatched(event *Event, handlerIDs []int64) error {
	event.bot = b

	b.mu.RLock()
	callbacks := make([]EventHandler, 0, len(handlerIDs))
	var errs []error
	for _, id := range handlerIDs {
		handler, ok := b.handlers[id]
		if !ok {
			errs = append(errs, fmt.Errorf("lsmsg: handler %d not found", id))
			continue
		}
		callbacks = append(callbacks, handler)
	}
	b.mu.RUnlock()

	for _, handler := range callbacks {
		if err := handler(event); err != nil {
			errs = append(errs, err)
		}
	}
	return errors.Join(errs...)
}

// dispatch runs all matching handlers for an event.
func (b *Bot) dispatch(event *Event) error {
	handlerIDs, err := b.matchHandlerIDs(event)
	if err != nil {
		return err
	}
	return b.dispatchMatched(event, handlerIDs)
}

// ServeHTTP implements http.Handler. It routes requests to the appropriate
// platform webhook handler based on the URL path.
func (b *Bot) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodPost {
		http.Error(w, "method not allowed", http.StatusMethodNotAllowed)
		return
	}

	switch r.URL.Path {
	case "/slack/events", "/slack":
		b.HandleSlackWebhook(w, r)
	case "/teams/events", "/teams":
		b.HandleTeamsWebhook(w, r)
	default:
		// Try to auto-detect based on headers.
		if r.Header.Get("X-Slack-Signature") != "" {
			b.HandleSlackWebhook(w, r)
			return
		}
		// Default to a generic handler that tries both.
		http.Error(w, "unknown webhook path", http.StatusNotFound)
	}
}

func (b *Bot) writeWebhookOutcome(ctx context.Context, w http.ResponseWriter, outcome *webhookOutcome) {
	if outcome == nil {
		http.Error(w, "failed to process webhook", http.StatusInternalServerError)
		return
	}

	switch outcome.Type {
	case "rejected":
		statusCode := outcome.StatusCode
		if statusCode == 0 {
			statusCode = http.StatusBadRequest
		}
		message := outcome.Error
		if message == "" {
			message = http.StatusText(statusCode)
		}
		http.Error(w, message, statusCode)
	case "challenge":
		w.Header().Set("Content-Type", "application/json")
		_ = json.NewEncoder(w).Encode(map[string]string{"challenge": outcome.Challenge})
	case "ignored":
		w.WriteHeader(http.StatusOK)
	case "dispatch":
		if outcome.Event == nil {
			http.Error(w, "failed to process webhook", http.StatusInternalServerError)
			return
		}
		b.dispatchAsync(ctx, outcome.Event, outcome.HandlerIDs, w)
	default:
		http.Error(w, "failed to process webhook", http.StatusInternalServerError)
	}
}

// HandleSlackWebhook handles an incoming Slack webhook HTTP request.
func (b *Bot) HandleSlackWebhook(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()

	body, err := io.ReadAll(io.LimitReader(r.Body, int64(b.config.MaxBodyBytes)))
	if err != nil {
		http.Error(w, "failed to read body", http.StatusBadRequest)
		return
	}
	defer r.Body.Close()

	signingSecret := ""
	if b.config.Slack != nil {
		signingSecret = b.config.Slack.SigningSecret
	}
	outcome, err := b.ffi.RegistryProcessSlackWebhook(
		b.registryHandle,
		body,
		r.Header.Get("Content-Type"),
		signingSecret,
		r.Header.Get("X-Slack-Request-Timestamp"),
		r.Header.Get("X-Slack-Signature"),
	)
	if err != nil {
		http.Error(w, "failed to process webhook", http.StatusInternalServerError)
		return
	}
	b.writeWebhookOutcome(ctx, w, outcome)
}

// HandleTeamsWebhook handles an incoming Teams webhook HTTP request.
func (b *Bot) HandleTeamsWebhook(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()

	body, err := io.ReadAll(io.LimitReader(r.Body, int64(b.config.MaxBodyBytes)))
	if err != nil {
		http.Error(w, "failed to read body", http.StatusBadRequest)
		return
	}
	defer r.Body.Close()

	outcome, err := b.ffi.RegistryProcessTeamsWebhook(b.registryHandle, body)
	if err != nil {
		http.Error(w, "failed to process webhook", http.StatusInternalServerError)
		return
	}
	b.writeWebhookOutcome(ctx, w, outcome)
}

func (b *Bot) dispatchAsync(_ context.Context, event *Event, handlerIDs []int64, w http.ResponseWriter) {
	// Acknowledge immediately, dispatch in background.
	if err := b.dispatchMatched(event, handlerIDs); err != nil {
		// Log but still return 200 to the platform.
		_ = err
	}
	w.WriteHeader(http.StatusOK)
}

// Invoke creates a LangGraph run and waits for its result.
func (e *Event) Invoke(agent string, opts ...InvokeOpt) (*RunResult, error) {
	if e.bot == nil {
		return nil, errors.New("lsmsg: event not associated with a bot")
	}
	if !e.bot.langGraphConfigured {
		return nil, errors.New("lsmsg: LangGraph not configured")
	}

	cfg := &invokeConfig{}
	for _, o := range opts {
		o(cfg)
	}

	threadID := e.bot.ffi.DeterministicThreadID(
		string(e.Platform.Name),
		e.WorkspaceID,
		e.ChannelID,
		e.ThreadID,
	)

	input := cfg.input
	if input == nil {
		input = map[string]any{
			"messages": []map[string]any{
				{"role": "user", "content": e.Text},
			},
		}
	}

	params := map[string]any{
		"agent":     agent,
		"thread_id": threadID,
	}
	if input != nil {
		params["input"] = input
	}
	if cfg.config != nil {
		params["config"] = cfg.config
	}
	if cfg.metadata != nil {
		params["metadata"] = cfg.metadata
	}

	paramsJSON, err := json.Marshal(params)
	if err != nil {
		return nil, fmt.Errorf("lsmsg: failed to marshal params: %w", err)
	}

	runID, err := e.bot.ffi.LangGraphCreateRun(e.bot.lgHandle, paramsJSON)
	if err != nil {
		return nil, fmt.Errorf("lsmsg: create run failed: %w", err)
	}

	result, err := e.bot.ffi.LangGraphWaitRun(e.bot.lgHandle, threadID, runID)
	if err != nil {
		return nil, fmt.Errorf("lsmsg: wait run failed: %w", err)
	}

	return result, nil
}

// Reply sends a text reply back to the originating platform channel/thread.
// For Slack this uses chat.postMessage; for Teams it uses the activity reply API.
func (e *Event) Reply(text string) error {
	if e.bot == nil {
		return errors.New("lsmsg: event not associated with a bot")
	}

	switch e.Platform.Name {
	case PlatformSlack:
		return e.replySlack(text)
	case PlatformTeams:
		return e.replyTeams(text)
	default:
		return fmt.Errorf("lsmsg: reply not implemented for platform %s", e.Platform.Name)
	}
}

func (e *Event) replySlack(text string) error {
	if e.bot.config.Slack == nil || e.bot.config.Slack.BotToken == "" {
		return errors.New("lsmsg: Slack bot token not configured")
	}

	payload := map[string]string{
		"channel":   e.ChannelID,
		"text":      text,
		"thread_ts": e.ThreadID,
	}
	body, _ := json.Marshal(payload)

	url := strings.TrimRight(e.bot.slackAPIBaseURL, "/") + "/api/chat.postMessage"
	req, err := http.NewRequest(http.MethodPost, url, nil)
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/json; charset=utf-8")
	req.Header.Set("Authorization", "Bearer "+e.bot.config.Slack.BotToken)
	req.Body = io.NopCloser(jsonReader(body))

	resp, err := e.bot.httpClient.Do(req)
	if err != nil {
		return fmt.Errorf("lsmsg: slack reply failed: %w", err)
	}
	defer resp.Body.Close()

	respBody, _ := io.ReadAll(resp.Body)
	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("lsmsg: slack reply returned %d: %s", resp.StatusCode, string(respBody))
	}

	var result struct {
		OK    bool   `json:"ok"`
		Error string `json:"error"`
	}
	if err := json.Unmarshal(respBody, &result); err != nil {
		return fmt.Errorf("lsmsg: slack reply returned invalid JSON: %w", err)
	}
	if !result.OK {
		if result.Error == "" {
			result.Error = "unknown_error"
		}
		return fmt.Errorf("lsmsg: slack reply failed: %s", result.Error)
	}
	return nil
}

func (e *Event) replyTeams(_ string) error {
	// Teams reply requires the service URL from the original activity,
	// which would be stored in the raw event. Placeholder for now.
	return errors.New("lsmsg: Teams reply not yet implemented")
}

// jsonReader wraps a byte slice as an io.Reader.
func jsonReader(b []byte) io.Reader {
	return &jsonBytesReader{data: b}
}

type jsonBytesReader struct {
	data []byte
	pos  int
}

func (r *jsonBytesReader) Read(p []byte) (int, error) {
	if r.pos >= len(r.data) {
		return 0, io.EOF
	}
	n := copy(p, r.data[r.pos:])
	r.pos += n
	return n, nil
}

File: sdks/typescript/src/bot.ts
import type { Request, Response, Router } from 'express';
import type {
  BotConfig,
  Event,
  EventHandler,
  EventKind,
  HandlerOptions,
  InvokeOptions,
  Platform,
  PlatformCapabilities,
  RunChunk,
  RunResult,
} from './types.js';

// The napi native module. Loaded lazily so tests can inject a mock.
let native: any;
try {
  // eslint-disable-next-line @typescript-eslint/no-require-imports
  native = require('@botmux/native');
} catch {
  // Not available — must be set via setNativeModule before creating a Bot.
  native = undefined;
}

/** @internal Replace the native module (used by tests). */
export function setNativeModule(mod: any): void {
  native = mod;
}

function assertNativeAvailable(): void {
  if (native) {
    return;
  }
  throw new Error(
    'botmux native bindings are unavailable. Install @botmux/native or inject a test double with setNativeModule().',
  );
}

/** @internal Convert snake_case napi event JSON to camelCase Event with methods. */
function toEvent(raw: Record<string, any>, bot: Bot): Event {
  const platform: PlatformCapabilities = {
    name: raw.platform?.name as Platform,
    ephemeral: raw.platform?.ephemeral ?? false,
    threads: raw.platform?.threads ?? false,
    reactions: raw.platform?.reactions ?? false,
    streaming: raw.platform?.streaming ?? false,
    modals: raw.platform?.modals ?? false,
    typingIndicator: raw.platform?.typing_indicator ?? false,
  };

  const internalThreadId = native.deterministicThreadId(
    platform.name,
    raw.workspace_id ?? '',
    raw.channel_id ?? '',
    raw.thread_id ?? '',
  );

  const event: Event = {
    kind: raw.kind as EventKind,
    platform,
    workspaceId: raw.workspace_id ?? '',
    channelId: raw.channel_id ?? '',
    threadId: raw.thread_id ?? '',
    messageId: raw.message_id ?? '',
    user: {
      id: raw.user?.id ?? '',
      name: raw.user?.name,
      email: raw.user?.email,
    },
    text: raw.text ?? '',
    command: raw.command,
    emoji: raw.emoji,
    rawEventType: raw.raw_event_type,
    raw: raw.raw,
    internalThreadId,

    async invoke(agent: string, options?: InvokeOptions): Promise<RunResult> {
      return bot._invoke(event, agent, options);
    },

    async stream(agent: string, options?: InvokeOptions): Promise<RunChunk[]> {
      return bot._stream(event, agent, options);
    },

    async reply(text: string): Promise<void> {
      return bot._reply(event, text);
    },
  };

  return event;
}

/**
 * Wraps a synchronous napi call in a Promise that yields to the event loop
 * via setImmediate before executing.
 */
function deferSync<T>(fn: () => T): Promise<T> {
  return new Promise((resolve, reject) => {
    setImmediate(() => {
      try {
        resolve(fn());
      } catch (err) {
        reject(err);
      }
    });
  });
}

interface RegisteredHandler {
  id: number;
  callback: EventHandler;
}

type RawBodyRequest = Request & { rawBody?: Buffer | string };

/**
 * The main Bot class. Provides an idiomatic Node.js API for registering
 * event handlers and processing webhooks from Slack, Teams, and other platforms.
 */
export class Bot {
  private readonly config: BotConfig;
  private readonly registry: InstanceType<typeof native.HandlerRegistry>;
  private readonly handlers: Map<number, RegisteredHandler> = new Map();
  constructor(config: BotConfig) {
    assertNativeAvailable();
    this.config = config;
    this.registry = new native.HandlerRegistry();
  }

  // ---------------------------------------------------------------------------
  // Handler registration
  // ---------------------------------------------------------------------------

  /** Register a handler for @-mention events. */
  mention(handler: EventHandler, options?: HandlerOptions): number {
    return this._register('mention', undefined, undefined, options, handler);
  }

  /** Register a handler for plain message events. */
  message(handler: EventHandler, options?: HandlerOptions): number {
    return this._register('message', undefined, undefined, options, handler);
  }

  /** Register a handler for a slash command. */
  command(
    name: string,
    handler: EventHandler,
    options?: HandlerOptions,
  ): number {
    return this._register('command', name, undefined, options, handler);
  }

  /** Register a handler for a specific emoji reaction. */
  reaction(
    emoji: string,
    handler: EventHandler,
    options?: HandlerOptions,
  ): number {
    return this._register('reaction', undefined, emoji, options, handler);
  }

  /** Register a handler for a raw platform event type or a broad event kind. */
  on(eventType: EventKind | string, handler: EventHandler, options?: HandlerOptions): number {
    if (
      eventType === 'message'
      || eventType === 'mention'
      || eventType === 'command'
      || eventType === 'reaction'
      || eventType === 'raw'
    ) {
      return this._register(eventType, undefined, undefined, options, handler);
    }
    return this._register('raw', undefined, undefined, options, handler, eventType);
  }

  /** Remove a previously registered handler. Returns true if it existed. */
  off(handlerId: number): boolean {
    this.handlers.delete(handlerId);
    return this.registry.unregister(handlerId);
  }

  // ---------------------------------------------------------------------------
  // Webhook handlers
  // ---------------------------------------------------------------------------

  /**
   * Express-compatible request handler for Slack webhooks.
   * Verifies the signature, parses the payload, and dispatches to matching handlers.
   */
  async handleSlackWebhook(req: Request, res: Response): Promise<void> {
    const signingSecret = this.config.slack?.signingSecret;
    if (!signingSecret) {
      res.status(500).json({ error: 'Slack signing secret not configured' });
      return;
    }

    let rawBody: string;
    try {
      rawBody = await getRawBody(req, Boolean(signingSecret));
    } catch (err) {
      const message = err instanceof Error ? err.message : 'Invalid request body';
      res.status(400).json({ error: message });
      return;
    }
    const timestamp = (req.headers['x-slack-request-timestamp'] as string) ?? '';
    const signature = (req.headers['x-slack-signature'] as string) ?? '';

    const contentType = (req.headers['content-type'] as string) ?? 'application/json';
    const parsed = await deferSync(() =>
      this.registry.processSlackWebhook(
        Buffer.from(rawBody),
        contentType,
        signingSecret,
        timestamp,
        signature,
      ),
    );

    if (parsed.type === 'rejected') {
      res.status(parsed.status_code ?? 400).json({ error: parsed.error ?? 'Rejected' });
      return;
    }

    if (parsed.type === 'challenge') {
      res.status(200).json({ challenge: parsed.challenge });
      return;
    }

    if (parsed.type === 'ignored' || !parsed.event) {
      res.status(200).json({ ok: true });
      return;
    }

    // Acknowledge immediately
    res.status(200).json({ ok: true });

    const event = toEvent(parsed.event, this);
    await this._dispatch(event, parsed.handler_ids as number[] | undefined);
  }

  /**
   * Express-compatible request handler for Teams webhooks.
   */
  async handleTeamsWebhook(req: Request, res: Response): Promise<void> {
    const body = await getRawBody(req as RawBodyRequest);
    const parsed = await deferSync(() =>
      this.registry.processTeamsWebhook(Buffer.from(body)),
    );

    if (parsed.type === 'rejected') {
      res.status(parsed.status_code ?? 400).json({ error: parsed.error ?? 'Rejected' });
      return;
    }

    if (parsed.type === 'ignored' || !parsed.event) {
      res.status(200).json({ ok: true });
      return;
    }

    res.status(200).json({ ok: true });

    const event = toEvent(parsed.event, this);
    await this._dispatch(event, parsed.handler_ids as number[] | undefined);
  }

  /**
   * Returns an Express Router with webhook routes mounted.
   * @param prefix - Optional route prefix (default: '/').
   */
  expressMiddleware(prefix?: string): Router {
    // Dynamic import to keep express optional
    // eslint-disable-next-line @typescript-eslint/no-require-imports
    const express = require('express');
    const router: Router = express.Router();
    const base = prefix ? prefix.replace(/\/+$/, '') : '';
    const slackBodyParser = express.raw({
      type: () => true,
      verify: (req: RawBodyRequest, _res: Response, buf: Buffer) => {
        req.rawBody = Buffer.from(buf);
      },
    });
    const teamsBodyParser = express.raw({
      type: () => true,
      verify: (req: RawBodyRequest, _res: Response, buf: Buffer) => {
        req.rawBody = Buffer.from(buf);
      },
    });

    router.post(`${base}/slack/events`, slackBodyParser, (req, res) => {
      this.handleSlackWebhook(req, res).catch((err) => {
        console.error('[botmux] Slack webhook error:', err);
        if (!res.headersSent) {
          res.status(500).json({ error: 'Internal error' });
        }
      });
    });

    router.post(`${base}/teams/events`, teamsBodyParser, (req, res) => {
      this.handleTeamsWebhook(req, res).catch((err) => {
        console.error('[botmux] Teams webhook error:', err);
        if (!res.headersSent) {
          res.status(500).json({ error: 'Internal error' });
        }
      });
    });

    return router;
  }

  /**
   * Start an Express server listening on the given port with webhook routes.
   */
  listen(port: number): void {
    // eslint-disable-next-line @typescript-eslint/no-require-imports
    const express = require('express');
    const app = express();
    app.use(this.expressMiddleware());
    app.listen(port, () => {
      console.log(`[botmux] Bot listening on port ${port}`);
    });
  }

  // ---------------------------------------------------------------------------
  // Internal: LangGraph integration (to be implemented with native fetch)
  // ---------------------------------------------------------------------------

  /** @internal */
  async _invoke(
    _event: Event,
    _agent: string,
    _options?: InvokeOptions,
  ): Promise<RunResult> {
    throw new Error(
      'LangGraph client not yet implemented in TypeScript SDK. Use fetch directly.',
    );
  }

  /** @internal */
  async _stream(
    _event: Event,
    _agent: string,
    _options?: InvokeOptions,
  ): Promise<RunChunk[]> {
    throw new Error(
      'LangGraph client not yet implemented in TypeScript SDK. Use fetch directly.',
    );
  }

  /** @internal */
  async _reply(event: Event, text: string): Promise<void> {
    // Platform-specific reply logic. Full implementation requires platform API
    // tokens and HTTP calls. For now, log the reply.
    console.log(
      `[botmux] Reply to ${event.platform.name}/${event.channelId}: ${text}`,
    );
  }

  // ---------------------------------------------------------------------------
  // Internal: registration and dispatch
  // ---------------------------------------------------------------------------

  /** @internal */
  private _register(
    eventKind: EventKind,
    command: string | undefined,
    emoji: string | undefined,
    options: HandlerOptions | undefined,
    callback: EventHandler,
    rawEventType?: string,
  ): number {
    const id: number = this.registry.register(
      eventKind,
      command ?? null,
      options?.pattern ?? null,
      emoji ?? null,
      options?.platform ?? null,
      rawEventType ?? null,
    );
    this.handlers.set(id, { id, callback });
    return id;
  }

  /** @internal */
  private async _dispatch(event: Event, matchedIds?: number[]): Promise<void> {
    // Convert event back to snake_case JSON for the napi registry matcher
    const eventJson = {
      kind: event.kind,
      platform: {
        name: event.platform.name,
        ephemeral: event.platform.ephemeral,
        threads: event.platform.threads,
        reactions: event.platform.reactions,
        streaming: event.platform.streaming,
        modals: event.platform.modals,
        typing_indicator: event.platform.typingIndicator,
      },
      workspace_id: event.workspaceId,
      channel_id: event.channelId,
      thread_id: event.threadId,
      message_id: event.messageId,
      user: event.user,
      text: event.text,
      command: event.command,
      emoji: event.emoji,
      raw_event_type: event.rawEventType,
      raw: event.raw ?? null,
    };

    const resolvedIds = matchedIds ?? await deferSync(() =>
      this.registry.matchEvent(eventJson),
    );

    const promises: Promise<void>[] = [];
    for (const id of resolvedIds) {
      const registered = this.handlers.get(id);
      if (registered) {
        promises.push(
          new Promise<void>((resolve) => resolve(registered.callback(event) as any)).catch((err) => {
            console.error(`[botmux] Handler ${id} threw:`, err);
          }),
        );
      }
    }

    await Promise.all(promises);
  }
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function extractText(result: any): string {
  const messages = result?.output?.messages ?? result?.messages;
  if (Array.isArray(messages) && messages.length > 0) {
    const last = messages[messages.length - 1];
    if (typeof last?.content === 'string') {
      return last.content;
    }
  }
  return '';
}

async function getRawBody(req: RawBodyRequest, requireOriginalBody = false): Promise<string> {
  if (Buffer.isBuffer(req.rawBody)) {
    return req.rawBody.toString('utf-8');
  }
  if (typeof req.rawBody === 'string') {
    return req.rawBody;
  }
  // If body is already parsed as a Buffer or string
  if (Buffer.isBuffer(req.body)) {
    return req.body.toString('utf-8');
  }
  if (typeof req.body === 'string') {
    return req.body;
  }
  if (req.body && typeof req.body === 'object') {
    if (requireOriginalBody) {
      throw new Error(
        'Slack signature verification requires the original raw request body. Use bot.expressMiddleware() or provide req.rawBody.',
      );
    }
    return JSON.stringify(req.body);
  }
  // Read from stream
  return new Promise<string>((resolve, reject) => {
    const chunks: Buffer[] = [];
    req.on('data', (chunk: Buffer) => chunks.push(chunk));
    req.on('end', () => resolve(Buffer.concat(chunks).toString('utf-8')));
    req.on('error', reject);
  });
}
